lahja/endpoint.py

```diff
@@ -43,68 +43,118 @@ def __init__(self,
         self._queues: Dict[Type[BaseEvent], List[asyncio.Queue]] = {}
-        self._running = False
         self._executor: Optional[ThreadPoolExecutor] = None
+        self._running = False
+        self._optional_internal_queue: Optional[asyncio.Queue] = None
+        self._optional_loop: Optional[asyncio.AbstractEventLoop] = None
+        self._optional_receiving_loop_running: Optional[asyncio.Event] = None
+        self._optional_internal_loop_running: Optional[asyncio.Event] = None
+
+    @property
+    def _loop(self) -> asyncio.AbstractEventLoop:
+        if self._optional_loop is None:
+            raise NoConnection("Need to connect first")
+        return self._optional_loop
+
+    @property
+    def _internal_queue(self) -> asyncio.Queue:
+        if self._optional_internal_queue is None:
+            raise NoConnection("Need to connect first")
+        return self._optional_internal_queue
+
+    @property
+    def _receiving_loop_running(self) -> asyncio.Event:
+        if self._optional_receiving_loop_running is None:
+            raise NoConnection("Need to connect first")
+        return self._optional_receiving_loop_running
 
     @property
-    def loop(self) -> asyncio.AbstractEventLoop:
-        if self._loop is None:
-            raise NoConnection("Need to call `connect()` first")
-        return self._loop
+    def _internal_loop_running(self) -> asyncio.Event:
+        if self._optional_internal_loop_running is None:
+            raise NoConnection("Need to connect first")
+        return self._optional_internal_loop_running
 
-    def connect(self, loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
+    def connect_no_wait(self, loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
         """
         Connect the :class:`~lahja.endpoint.Endpoint` to the :class:`~lahja.eventbus.EventBus`
         instance that created this endpoint.
         """
-        if (loop is None):
+        if loop is None:
             loop = asyncio.get_event_loop()
 
-        self._loop = loop
-        asyncio.ensure_future(self._try_connect(loop), loop=loop)
+        self._optional_loop = loop
+        self._optional_internal_loop_running = asyncio.Event(loop=self._loop)
+        self._optional_receiving_loop_running = asyncio.Event(loop=self._loop)
+        self._optional_internal_queue = asyncio.Queue()
 
-    async def _try_connect(self, loop: asyncio.AbstractEventLoop) -> None:
-        # We need to handle exceptions here to not get `Task exception was never retrieved`
-        # errors in case the `connect()` fails (e.g. because the remote process is shutting down)
-        try:
-            await asyncio.ensure_future(self._connect(), loop=loop)
-        except Exception as exc:
-            raise Exception("Exception from Endpoint.connect()") from exc
+        # Using `gather` (over e.g. `wait` or plain `ensure_future`) ensures that the inner futures
```

Comment:
That's quite interesting.  Wonder if we can use this elsewhere?

Reply:
Yes probably, it's one of the many surprises of Python asyncio ;)

https://docs.python.org/3/library/asyncio-task.html#asyncio.gather

>If `gather()` is cancelled, all submitted awaitables (that have not completed yet) are also cancelled.

Also interesting:

>Unlike `wait_for()`, `wait()` does not cancel the futures when a timeout occurs.